react/features/notifications/middleware.js

```diff
@@ -13,29 +13,104 @@ import {
 import { MiddlewareRegistry, StateListenerRegistry } from '../base/redux';
 import { PARTICIPANTS_PANE_OPEN } from '../participants-pane/actionTypes';
 
+import {
+    CLEAR_NOTIFICATIONS,
+    HIDE_NOTIFICATION,
+    SHOW_NOTIFICATION
+} from './actionTypes';
 import {
     clearNotifications,
-    hideRaiseHandNotifications,
+    hideNotification,
     showNotification,
     showParticipantJoinedNotification,
     showParticipantLeftNotification
 } from './actions';
-import { NOTIFICATION_TIMEOUT_TYPE } from './constants';
-import { joinLeaveNotificationsDisabled } from './functions';
+import {
+    NOTIFICATION_TIMEOUT_TYPE,
+    RAISE_HAND_NOTIFICATION_ID
+} from './constants';
+import { areThereNotifications, joinLeaveNotificationsDisabled } from './functions';
 
 /**
  * Middleware that captures actions to display notifications.
  *
  * @param {Store} store - The redux store.
  * @returns {Function}
  */
+
+const timers = new Map();
+
+const createTimeoutId = (notification, dispatch) => {
+    const {
+        timeout,
+        uid
+    } = notification;
+    const timerID = setTimeout(() => {
+        dispatch(hideNotification(uid));
+    }, timeout);
+
+    timers.set(uid, timerID);
+};
+
+const getNotifications = state => {
+    const _visible = areThereNotifications(state);
+    const { notifications } = state['features/notifications'];
+
+    return _visible ? notifications : [];
+};
+
 MiddlewareRegistry.register(store => next => action => {
+
+    const { dispatch, getState } = store;
+    const state = getState();
+
     switch (action.type) {
+    case CLEAR_NOTIFICATIONS: {
+        if (navigator.product !== 'ReactNative') {
+            const _notifications = getNotifications(state);
+
+            for (const notification of _notifications) {
+                if (timers.has(notification.uid)) {
+                    const timeout = timers.get(notification.uid);
+
+                    clearTimeout(timeout);
+                    timers.delete(notification.uid);
+                }
+            }
+            timers.clear();
+        }
+        break;
+    }
+    case SHOW_NOTIFICATION: {
+        if (navigator.product !== 'ReactNative') {
+            const _notifications = getNotifications(state);
+
+            for (const notification of _notifications) {
+                if (timers.has(notification.uid)) {
+
+                    const timer = timers.get(action.uid);
+
+                    clearTimeout(timer);
+                    timers.delete(notification.uid);
+                    createTimeoutId(notification, dispatch);
```

Comment:
This way you are rearming all timers every time a new notification (related or not) comes in. I think you need to just operate on the notification attached to the action instead of iterating through them all.

For example: if there is a notification related to you being noisy in the array and someone raises their hand, you don't want to rearm any times. 

The SHOW_NOTIFICATION action has a `uid` field which you can use for this.